File: grooves-core/src/main/java/com/github/rahulsom/grooves/functions/EventVersioner.java
package com.github.rahulsom.grooves.functions;

import com.github.rahulsom.grooves.logging.Trace;

/**
 * Provides the version of the event.
 */
@FunctionalInterface
public interface EventVersioner<EventT, VersionOrTimestampT> {
    /**
     * Extracts the version or timestamp from the given event.
     *
     * @param event the event to get the version/timestamp from
     * @return the version or timestamp of the event
     */
    @Trace
    VersionOrTimestampT invoke(EventT event);
}

File: grooves-core/src/main/kotlin/com/github/rahulsom/grooves/functions/EventVersioner.kt
package com.github.rahulsom.grooves.functions

import com.github.rahulsom.grooves.logging.Trace

/**
 * Provides the version of the event.
 */
interface EventVersioner<Event, VersionOrTimestamp> {
    @Trace
    fun invoke(event: Event): VersionOrTimestamp
}